src/metabase/models/serialization/base.clj

```diff
@@ -0,0 +1,171 @@
+
+    Default implementation:
+    - Uses `\"$ENTITY_ID.yaml\"` or `\"$IDENTITY_HASH.yaml\"` as the filename.
+    - Returns the entity as a vanilla map with `:id` and any `:foo_at` fields removed.
+    - Adds the field `:type \"Collection\"` (etc.) to specify the model.
+
+    That suffices for a few simple entities, but most entities will need to override this.
+    They should follow the pattern of dropping timestamps and numeric database IDs, and including the `:type`.")
+
+  ;;; Deserialization side.
+  (deserialize-file
+    [this new-map local]
+    "Given the model, the file contents as an EDN map, and any corresponding entity already existing in the database,
+    this performs whatever steps are necessary to update or insert the incoming entity.
+    Called for each **file**, which usually means for each entity. For single-file output like for settings, this is
+    called once for the whole file.
+    Defaults to [[deserialize-file-plus]] with [[identity]] as the inner transformation.")
+
+  (deserialize-upsert
+    [old-entity new-map]
+    "Given the original entity and the new deserialized map (not an IModel yet), upsert appdb to merge in the updated
+    entity.
+    Defaults to a naive update by primary key, using just the new map's values.")
+
+  (deserialize-insert
+    [this new-map]
+    "Given the model (eg. Collection) and the new deserialized map (not an IModel), insert this new record into
+    the appdb.
+    Defaults to a straight db/insert! of this new map."))
+
+(defn- default-serialize-all [model user-or-nil]
+  (eduction (map serialize-one) (serialize-query model user-or-nil)))
+
+(defn- default-query [model _]
+  (db/select-reducible model))
+
+(defn serialize-one-plus
+  "Helper for applying the usual serialization steps to a single entity, followed by a user-supplied function with any
+  specific logic the entity needs."
+  [transform named]
+  (fn [entity]
+    [(format "%s%s%s%s.yaml"
+             (name entity)
+             File/separatorChar
+             (or (:entity_id entity)
+                 (serdes.hash/identity-hash entity))
+             (when named
+               (str "+" (named entity))))
+     (-> (into {} entity)
+         (dissoc (models/primary-key entity))
+         (assoc :serdes_type (name entity))
+         transform)]))
+
+(defn- default-upsert [old-entity new-map]
+  (db/update! (symbol (name old-entity)) (get old-entity (models/primary-key old-entity)) new-map)
+  #_(prn "upsert" old-entity new-map))
+
+(defn- default-insert [model new-map]
+  (db/simple-insert! model new-map)
+  #_(prn "insert" new-map))
+
+(defn deserialize-file-plus
+  "Given a function, this returns a function suitable for use as [[deserialize-file]]. The provided function is applied
+  to the EDN map read from the YAML file, and then either [[deserialize-upsert]] or [[deserialize-insert]] is called
+  depending on whether a corresponding entity was found in the database."
+  [f]
```

Comment:
A few thoughts: 

- I would name `f` something like `xform` or `contents-xform` or better yet `contents-transform-fn` to make it clearer that it's a function used to transform the file contents. Personally I am of the you-can't-have-variable-names-that-are-too-long school. I think the longer names make the code easier to read
- Since this is a function that given a transform function returns a `deserialize-file` function I would name this function something like `deserialize-file-fn` or `deserialize-file-with-transform-fn`. Sticking `-fn` on the end makes it clear that it returns a `-fn`. I don't know where you got `-plus` from but NGL it confused me a lot when I first saw it

Reply:
The new factoring is both more transparent and better documented.